src/base/lib/Convex/BuildTx.hs

```diff
@@ -327,7 +338,57 @@ addWithdrawalWithTxBody :: (MonadBuildTx era m, C.IsShelleyBasedEra era) => C.St
 addWithdrawalWithTxBody address amount f =
   addTxBuilder (TxBuilder $ \body -> over (L.txWithdrawals . L._TxWithdrawals) ((address, C.quantityToLovelace amount, C.BuildTxWith $ f body) :))
 
+{-| Add a stake witness to the transaction.
+TODO We should probably remove this as the `addStakeScriptWitness` is more useful.
+-}
+addStakeWitness ::
+  ( MonadBuildTx era m
+  , C.IsShelleyBasedEra era
+  )
+  => C.StakeCredential
+  -> C.Witness C.WitCtxStake era
+  -> m ()
+addStakeWitness credential witness =
+  addBtx (over (L.txCertificates . L._TxCertificates . _2) ((:) (credential, witness)))
+
+{-| Add a stake script witness to the transaction.
+-}
+addStakeScriptWitness ::
+  ( MonadBuildTx era m
+  , Plutus.ToData redeemer
+  , C.IsShelleyBasedEra era
+  , C.IsPlutusScriptLanguage lang
+  , C.HasScriptLanguageInEra lang era
+  )
+  => C.StakeCredential
+  -> C.PlutusScript lang
+  -> redeemer
+  -> m ()
+addStakeScriptWitness credential script redeemer = do
+  let scriptWitness = buildScriptWitness script C.NoScriptDatumForStake redeemer
+  let witness = C.ScriptWitness C.ScriptWitnessForStakeAddr scriptWitness
+  addBtx (over (L.txCertificates . L._TxCertificates . _2) ((:) (credential, witness)))
+
+{-| Add a stake script reference witness to the transaction.
+-}
+addStakeScriptWitnessRef ::
+  ( MonadBuildTx era m
+  , Plutus.ToData redeemer
+  , C.IsShelleyBasedEra era
+  , C.HasScriptLanguageInEra lang era
+  )
+  => C.StakeCredential
+  -> C.TxIn
+  -> C.PlutusScriptVersion lang
+  -> redeemer
+  -> m ()
+addStakeScriptWitnessRef credential txIn plutusScriptVersion redeemer = do
+  let scriptWitness = buildRefScriptWitness txIn plutusScriptVersion C.NoScriptDatumForStake redeemer
+  let witness = C.ScriptWitness C.ScriptWitnessForStakeAddr scriptWitness
+  addBtx (over (L.txCertificates . L._TxCertificates . _2) ((:) (credential, witness)))
+
 {- | Like @addStakeWitness@ but uses a function that takes a @TxBody@ to build the witness.
+TODO Give an example of why this is useful. We should just remove it.
```

Comment:
This is useful if the redeemer needs to know the final set of inputs of the transaction. For example if we want to pass the index of a specific input to the redeemer.

Reply:
Ah cool! I'll document this in another PR